test/test_state_machine_enum.rb

```diff
@@ -2,12 +2,71 @@
 
 require "test_helper"
 
-class TestStateMachineEnum < Minitest::Test
-  def test_that_it_has_a_version_number
-    refute_nil ::StateMachineEnum::VERSION
+class TestStateMachineEnum < ActiveSupport::TestCase
+  class User < ActiveRecord::Base
+    include StateMachineEnum
+
+    state_machine_enum :state do |s|
+      s.permit_transition(:registered, :active)
+      s.permit_transition(:active, :banned)
+      s.permit_transition(:banned, :active)
+      s.permit_transition(:active, :deleted)
+    end
+  end
+
+  def test_state_transitions
+    u = User.create!
+
+    assert u.active!
+    assert u.banned!
+    assert u.active!
+    assert u.deleted!
+  end
+
+  def test_callbacks
+    u = User.create!
+
+    assert u.active!
+    error = assert_raises(ActiveRecord::RecordInvalid) { u.registered! }
+    assert_equal("Validation failed: State Invalid transition from active to registered", error.message)
+
+    error = assert_raises(ArgumentError) { u.update!(state: "missing") }
+    assert_equal("'missing' is not a valid state", error.message)
+  end
+
+  def test_ensure_state_may_transition_to!
+    u = User.create!(state: "active")
+
+    assert_nil u.ensure_state_may_transition_to!(:banned)
+
+    assert_raises(StateMachineEnum::InvalidState) { u.ensure_state_may_transition_to!("active") }
+    assert_raises(StateMachineEnum::InvalidState) { u.ensure_state_may_transition_to!(:active) }
+
+    # TODO: This method will not actually check if state can transition to the target state, it only verifies that this is not a current state.
```

Comment:
We can change that behavior I think. Should we rename this to `ensure_state_may_transition_to_or_is!`? and then do the actual check that it is permitted to transition?

Reply:
Name is fine, but the fact that we don't really check that state can transition was an issue. I have fixed that.